boot/freeldr/freeldr/arch/i386/pc/machpc.c

```diff
@@ -177,6 +177,63 @@ PcGetHarddiskConfigurationData(UCHAR DriveNumber, ULONG* pSize)
     return PartialResourceList;
 }
 
+static
+VOID
+DetectDockingStation(PCONFIGURATION_COMPONENT_DATA BusKey)
+{
+    PCM_PARTIAL_RESOURCE_LIST PartialResourceList;
+    PCM_PARTIAL_RESOURCE_DESCRIPTOR PartialDescriptor;
+    PCONFIGURATION_COMPONENT_DATA ControllerKey;
+    ULONG Size, Result;
+    PDOCKING_STATE_INFORMATION DockingState;
+
+    Result = PnpBiosGetDockStationInformation(DiskReadBuffer);
+
+    /* Build full device descriptor */
+    Size = sizeof(CM_PARTIAL_RESOURCE_LIST) + sizeof(DOCKING_STATE_INFORMATION);
+
+    if (Result == 0)
+    {
+        /* FIXME: Add more device specific data */
+        ERR("FIXME: System docked\n");
+    }
+
+    PartialResourceList = FrLdrHeapAlloc(Size, TAG_HW_RESOURCE_LIST);
+    if (PartialResourceList == NULL)
+    {
+        ERR("Failed to allocate resource descriptor\n");
+        return;
+    }
+
+    /* Initialize resource descriptor */
+    RtlZeroMemory(PartialResourceList, Size);
+    PartialResourceList->Version = 0;
+    PartialResourceList->Revision = 0;
+    PartialResourceList->Count = 1;
+
+    /* Set device specific data */
+    PartialDescriptor = &PartialResourceList->PartialDescriptors[0];
+    PartialDescriptor->Type = CmResourceTypeDeviceSpecific;
+    PartialDescriptor->ShareDisposition = CmResourceShareUndetermined;
+    PartialDescriptor->Flags = 0;
+    PartialDescriptor->u.DeviceSpecificData.DataSize = sizeof(DOCKING_STATE_INFORMATION);
+
+    DockingState = (PDOCKING_STATE_INFORMATION)&PartialResourceList->PartialDescriptors[1];
+    DockingState->ReturnCode = Result;
+
+    /* Create controller key */
+    FldrCreateComponentKey(BusKey,
+                           AdapterClass, // FIXME: Is this correct?
```

Comment:
What would help you in order to determine whether this is correct or not?

Reply:
I need a tool to view the ARC tree because configuration class is not exposed to the registry. By the way, VPC 2007 supports PNP BIOS calls.

Reply:
It cannot be "AdapterClass" since the docking information is within the "MultifunctionAdapter".
It looks to be an information-type thing like the RealModeIrqRoutingTable, i.e. a "PeripheralClass":
https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/freeldr/arch/i386/hwpci.c;hb=ff3d1b7bb1644aa629dfb9a149956825f8e6502c#l160

Reply:
You can add some extra DPRINTs in these functions of our kernel:
https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/config/cmconfig.c;hb=ff3d1b7bb1644aa629dfb9a149956825f8e6502c#l38
and try booting Windows 2003 on VPC but with this ROS modified kernel. This could allow you to view the ARC tree the NTLDR passes to the kernel.